src/resources/secret.ts

```diff
@@ -0,0 +1,80 @@
+// Copyright 2021, Nitric Technologies Pty Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import {
+  Action,
+  Resource,
+  ResourceType,
+  ResourceDeclareRequest,
+  ResourceDeclareResponse,
+} from '@nitric/api/proto/resource/v1/resource_pb';
+import resourceClient from './client';
+import { secrets, Secret } from '../api/secrets';
+import { ActionsList, make, Resource as Base } from './common';
+import { fromGrpcError } from '../api/errors';
+
+type SecretPermission = 'put' | 'access';
+
+const everything: SecretPermission[] = ['put', 'access'];
+
+/**
+ * Cloud secret resource for secret storage
+ */
+class SecretResource extends Base<SecretPermission> {
+  protected async register(): Promise<Resource> {
+    const req = new ResourceDeclareRequest();
+    const resource = new Resource();
+    resource.setName(this.name);
+    resource.setType(ResourceType.SECRET);
+    
+    req.setResource(resource);
+
+    return new Promise<Resource>((resolve, reject) => {
+      resourceClient.declare(
+        req,
+        (error, response: ResourceDeclareResponse) => {
+          if (error) {
+            reject(fromGrpcError(error));
+          } else {
+            resolve(resource);
+          }
+        }
+      )
+    })
+  }
+
+  protected permsToActions(...perms: SecretPermission[]): ActionsList {
+      return perms.reduce((actions, perm) => {
+        switch(perm) {
+          case 'put':
```

Comment:
or "update"? for("put") sounds a bit odd

Reply:
@asalkeld Using 'put' at the moment as the secret stores we're integration with are all append, so we're actually just adding a new secret value which will be rotated in at the `latest` secret value. I think update is fine as well, but in some ways implies that the old secret value will no longer exist, but it does still hang around.